src/index.ts

```diff
@@ -106,7 +106,37 @@ async function activate(
 
         console.log('loaded horus magic command extension');
 
-        // Use new message introduced in https://github.com/jupyterlab/jupyterlab/issues/10259
+        notebookTracker.activeCellChanged.connect(() => {
+
+          var track_change: boolean = false
+          notebookTracker.activeCell.editor.model.value.changed.connect(() => {
+            // user is typing
+            if ( notebookTracker.activeCell.editor.model.value.text.startsWith( "!pip" )) {
+              track_change = true
+              INotification.error("Do not use !pip install to handle dependencies, you cannot guarantee reproducibility. Use %horus magic commands!");
+            }
+          })
+
+          if ( (notebookTracker.activeCell.editor.model.value.text.startsWith( "!pip install" )) && ( track_change == false) ) {
+            INotification.error("Do not use !pip install to handle dependencies, you cannot guarantee reproducibility. Use `%horus clean` magic command to convert your cells.");
```

Comment:
About the message: I think it would be good to identify where does it come from (i.e. that it is this plugin). And maybe discourage direct `!pip` (and `%pip`), not only the `install` command?

In order to keep it short, maybe we can leave the evangelism for the docs :smiley:, so maybe take out *you cannot guaarantee reproducibility*.

All in all, what about something like: "Manage dependencies: please do not use direct `pip` commands. Use `horus` and/or the UI plugin instead."

Reply:
Great point @codificat :) I will adjust that as well!